lib/active_fedora/associations/directly_contains_one_association.rb

```diff
@@ -0,0 +1,118 @@
+module ActiveFedora
+  module Associations
+    # Filters a DirectContainer relationship, returning the first item that matches the given :type
+    class DirectlyContainsOneAssociation < SingularAssociation #:nodoc:
+
+      # Finds objects contained by the container predicate (either the configured has_member_relation or ldp:contains)
+      # TODO: Refactor this to use solr (for efficiency) instead of parsing the RDF graph.  Requires indexing ActiveFedora::File objects into solr, including their RDF.type and, if possible, the id of their container
+      def find_target
+        # filtered_objects = container_association_proxy.to_a.select { |o| o.metadata_node.type.include?(options[:type]) }
+        query_node = if container_predicate = options[:has_member_relation]
+                       owner
+                     else
+                       container_predicate = ::RDF::Vocab::LDP.contains
+                       container_association.container  # Use the :through association's container
+                     end
+
+        contained_uris = query_node.resource.query(predicate: container_predicate).map { |r| r.object.to_s }
+        contained_objects = contained_uris.map { |object_uri| klass.find(klass.uri_to_id(object_uri)) }
```

Comment:
A more efficient way to do this is:
```ruby
return contained_uris.find do |object_uri|
  contained_object = klass.find(klass.uri_to_id(object_uri))
  contained_object.metadata_node.type.include?(options[:type])
end
```
Then it only needs to load objects until it finds one that matches, rather than load all the objects.